src/imu/mpu9255/mpu9255.cpp

```diff
@@ -0,0 +1,150 @@
+#include "mpu9255.hpp"
+
+namespace imu {
+
+auto Mpu9255::Init(void) noexcept -> types::DriverStatus {
+  CreateSensorPointer();
+
+  SetInitConfigMPU9255();
+
+  if (InitAllSensors()) {
+    initialized_ = true;
+    return types::DriverStatus::OK;
+  }
+  return types::DriverStatus::HAL_ERROR;
+}
+
+auto Mpu9255::CreateSensorPointer(void) noexcept -> void {
+  if (gyroscope_ == NULL) {
+    gyroscope_ = std::make_unique<imu::Gyroscope>(i2c_handler_);
+  }
+  if (accelerometer_ == NULL) {
+    accelerometer_ = std::make_unique<imu::Accelerometer>(i2c_handler_);
+  }
+  if (magnetometer_ == NULL) {
+    magnetometer_ = std::make_unique<imu::Magnetometer>(i2c_handler_);
+  }
+  if (temperature_ == NULL) {
+    temperature_ = std::make_unique<imu::Temperature>(i2c_handler_);
+  }
+}
+
+auto Mpu9255::SetInitConfigMPU9255(void) noexcept -> void {
+  SetMPU9255Register(INT_PIN_CFG, 0x22);
+  SetMPU9255Register(INT_ENABLE, 0x01);
+}
+
+auto Mpu9255::SetMPU9255Register(const std::uint8_t register_, const std::uint8_t register_value) noexcept -> void {
+  i2c_handler_->Write(MPU9255_ADDRESS, {register_, register_value});
+}
+
+auto Mpu9255::InitAllSensors(void) noexcept -> bool {
+  if (gyroscope_->Init(MPU9255_ADDRESS) == types::DriverStatus::OK &&
+      accelerometer_->Init(MPU9255_ADDRESS) == types::DriverStatus::OK &&
+      magnetometer_->Init(AK8963_ADDRESS) == types::DriverStatus::OK &&
+      temperature_->Init(MPU9255_ADDRESS) == types::DriverStatus::OK) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+auto Mpu9255::Update(void) noexcept -> types::DriverStatus {
+  if (!IsInitialized())
+    return types::DriverStatus::HAL_ERROR;
+
+  if (UpdateAllSensors()) {
+    initialized_ = true;
```

Comment:
You mix levels of abstractions here.Here you use your private member variable initialized_ and 3 lines above you use a wrapper method

Reply:
Absolutely correct. I added SetToInitialized() to keep level of abstraction.